KAFKA-12495: Exponential backoff retry to prevent rebalance storms when worker joins after revoking rebalance

vamossagar12 opened the pull request:
Currently, the Incremental rebalance protocol does not allow a subsequent revoking rebalance when a worker joins right after one. This can lead to imbalance in assignments. See [KAFKA-12495](https://issues.apache.org/jira/browse/KAFKA-12495) for more details.

This PR aims to fix the above. Note that there already exists another PR: https://github.com/apache/kafka/pull/10367 to fix this. The main difference between the 2 approaches is that this one introduces an exponential backoff delay between 2 successive revoking rebalances. This is to dis-allow rebalance storms and still not wait for entire scheduled rebalance delay.

Notable changes in this PR =>

1) Allow successive revoking rebalances from the same leader.
2) If there are successive rebalances, then we would introduce a delay between the 2 based on exponential backoff policy. This delay would be used using the existing delay/scheduledRebalance mechanism.
3) The exponential backoff clock would be reset when the cluster reaches a balanced load i.e, there aren't successive revoking rebalances.

Commit:
KAFKA-12495: Initial commit to get a balanced assignment and allow consecutive rebalance

Commit:
Some local changes

Commit:
KAFKA-12495: Adding exponential backoff delay when worker joins after a revoking rebalance before a revoking rebalance can be triggered

vamossagar12 commented:
@C0urante , @showuon , @yashmayya I have created this PR as per the discussions on the ticket. Plz review whenever you get the chance. Thanks!

C0urante commented:
Thanks @vamossagar12. I haven't looked too closely at the code yet but have a couple high-level thoughts:

1. We should not introduce a backoff delay the first time a new worker joins the cluster. We should only introduce the delay in between consecutive rounds with load-balancing revocation.
2. We should reset the backoff delay as soon as we're able to complete a round of rebalancing that does not require any load-balancing revocations. IOW, as soon as everyone rejoins the cluster and we don't see any imbalances that need to be rectified by revoking connectors/tasks from one set of workers in one round, then re-allocating them to a different set of workers in a follow-up round. If we've just performed a round of revocation, and everyone has now joined the cluster and we're going to allocate those revoked connectors/tasks to new workers, as long as that allocation would lead to a balanced assignment, we can reset the clock.

vamossagar12 commented:
In this PR, similar to Luke's PR, I have added the condition to do a revoking rebalance consecutively. The only difference is that whether revocation would happen or not is decided by the time vis-a-vis the next scheduled revoking rebalance as per exponential backoff. I could notice the flip-side of that as I needed to add a delay in some of the tests to get a revoking rebalance if the workers joined before the next scheduled rebalance. 

I think what you are saying makes sense. Let me give it a try. Thanks!

Commit:
Adding revokedInPrevious state to exponential backoff delay only when there are successive revoking rebalabces

vamossagar12 commented:
@C0urante , i made the changes based on my understanding of your suggestions. Plz review whenever you get the chance. Thanks!

vamossagar12 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:
```diff
@@ -65,26 +67,32 @@ public class IncrementalCooperativeAssignor implements ConnectAssignor {
         this.previousGenerationId = -1;
         this.previousMembers = Collections.emptySet();
+        this.numSuccessiveRevokingRebalances = 0;
+        this.consecutiveRevokingRebalancesBackoff = new ExponentialBackoff(10, 20, maxDelay, 0.2);
```
@C0urante needed your suggestion on these configs. Should we expose the initial interval parameter so that users can set it? Also, the exp base(second parameter) I believe I have set it to a slightly high value which goes upto 4 seconds in the second attempt and about a minute in the third attempt and so on.

C0urante replied:
> Should we expose the initial interval parameter so that users can set it?

I don't think so; that would require a KIP to introduce a new user-facing configuration property.

> Also, the exp base(second parameter) I believe I have set it to a slightly high value which goes upto 4 seconds in the second attempt and about a minute in the third attempt and so on.

I think this is alright 👍
I don't see why we need jitter, though, and I'm hesitant to add it since it can lead to rebalance delays that exceed the `scheduled.rebalance.max.delay.ms` parameter. Is there a reason you opted for jitter here?

Also, it's worth noting that the behavior here is to always use a backoff of 10 ms if the `scheduled.rebalance.max.delay.ms` property is set to 10 or less. I'm not sure this is optimal; people may set the property to zero if they want to disable delayed rebalances altogether, and with this change, we'd be reintroducing delays with no way to disable them. Maybe we should:
- Set the initial interval to zero
- Check if the result of invoking `backoff` is zero, and if it is, skip the delay altogether and perform the revocation immediately

C0urante replied:
Oops, turns out we can't set the initial interval to `0` with the `ExponentialBackoff` class as that causes the backoff to be zero for every attempt. Hmm... this may involve a little bit of special handling on our end to accomplish.

vamossagar12 replied:
yeah i recall now that even i had tested it with 0 initial interval and it was always returning 0. Let me try to incorporate this .

vamossagar12 replied:
I added some logic to set delay to 0 when maxDelay is 0 in which case we would always revoke. I bumped up the multiplier from 20-> 30 as well since initial interval is reduced now.

showuon commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:
```diff
@@ -285,18 +292,11 @@ ClusterAssignment performTaskAssignment(
 
         handleLostAssignments(lostAssignments, newSubmissions, completeWorkerAssignment);
 
+        log.debug("Can leader revoke tasks in this assignment? {} (delay: {})", delay == 0, delay);
```
Since we don't have `canRevoke` now, I think we should rephrase this sentence. ex:
```java
if (delay > 0)  
    log.debug("Delaying {}ms for revoking tasks.");
```
WDYT?

vamossagar12 replied:
Makes sense. Would make the change

vamossagar12 replied:
This is done.

showuon commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:
```diff
@@ -584,9 +606,18 @@ private Map<String, ConnectorsAndTasks> performTaskRevocation(ConnectorsAndTasks
-        int newWorkersNum = totalWorkersNum - existingWorkersNum;
+        int newWorkersNum = totalWorkersNum - existingWorkersNum; // <total workers - workers with non-0 load>
+        // if existingWorkersNum == 0, no workers have any load i.e all workers are new.
+        // newWorkersNum = 0 when all workers have non-zero load.
```
newWorkersNum =[=] 0 ?

vamossagar12 replied:
I need to remove the comments. It was only for my understanding

vamossagar12 replied:
Removed.

showuon commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:
```diff
@@ -307,9 +307,31 @@ ClusterAssignment performTaskAssignment(
+                numSuccessiveRevokingRebalances++;
+                long processAfter = consecutiveRevokingRebalancesBackoff.backoff(numSuccessiveRevokingRebalances);
+                log.debug("Consecutive revoking rebalances observed. Need to wait for {} ms", processAfter);
+                time.sleep(processAfter); // Is it a good idea to sleep?
```
No, we can't sleep during the rebalance process. If there are some new tasks/connectors needed to distribute to workers, it'll be delayed for `processAfter` ms. I think we should use the existing `delay`/`scheduledRebalance` mechanism to delay the revocation.

vamossagar12 replied:
yeah that's what i thought as well. Using `delay/scheduledRebalance` might be a good idea.

vamossagar12 replied:
I have made the changes to use the existing delay/scheduledRebalance mechanism to delay the revocation.

showuon reviewed (commented):
Have a look at the non-test code, left some comments. And I agree with Chris that we should not introduce exponential backoff for 1st revocation, and should reset it ASAP we don't need it. Thank you for working on this.

vamossagar12 commented:
> Have a look at the non-test code, left some comments. And I agree with Chris that we should not introduce exponential backoff for 1st revocation, and should reset it ASAP we don't need it. Thank you for working on this.

Thanks @showuon . I removed the change to revoke the first time around. The delay is now introduced only when there are successive revoking rebalances.

Commit:
Removing time.sleep during successive rebalances and instead using the existing scheduled rebalanceDelay mechanism to introduce exponential backoffs during successive revoking rebalances

vamossagar12 commented:
Also, I have made the changes to use the existing delay/scheduledRebalance mechanism to delay the revocation. cc @showuon , @C0urante

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:
```diff
@@ -307,9 +309,31 @@ ClusterAssignment performTaskAssignment(
+            // backoff delay to prevent rebalance storms.
+            if (revokedInPrevious && !toExplicitlyRevoke.isEmpty()) {
+                numSuccessiveRevokingRebalances++;
+                delay = (int) consecutiveRevokingRebalancesBackoff.backoff(numSuccessiveRevokingRebalances);
```
We allow values up to `Integer.MAX_VALUE` for the `scheduled.rebalance.max.delay.ms` property. If we allow for jitter with our exponential backoff, this cast to `int` may lead to overflow when `scheduled.rebalance.max.delay.ms` has very large values.

vamossagar12 replied:
That's a good point and I didn't think about it. I used jitter to add some randomness but as you pointed out, it may not be needed. Would remove it.

vamossagar12 replied:
Removed.

C0urante commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java:
```diff
@@ -118,6 +119,93 @@ public void testTaskAssignmentWhenWorkerJoins() {
+        assertTaskAllocations(12);
+        assertBalancedAndCompleteAllocation();
+        // Flag should not be set
+        assertFalse(assignor.revokedInPrevious);
```
Why are we making assertions against this internal detail? Shouldn't the logic introduced in this PR be visible already by making assertions on connector/task allocations and the delay in each assignment?

vamossagar12 replied:
Removed.

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:
```diff
@@ -307,9 +309,31 @@ ClusterAssignment performTaskAssignment(
+            }
         } else {
-            canRevoke = delay == 0;
+            log.debug("Connector and task to revoke assignments: {}", toRevoke);
```
Why only log this in the `else` branch? Isn't it applicable in both cases?

vamossagar12 replied:
Added relevant logs..

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:
```diff
@@ -307,9 +309,31 @@ ClusterAssignment performTaskAssignment(
+                scheduledRebalance = time.milliseconds() + delay;
+            } else if (!toExplicitlyRevoke.isEmpty()) {
+                // We had a revocation in this round but not in the previous round. Let's store that state.
+                log.debug("Revoking rebalance. Setting the revokedInPrevious flag to true");
```
The name of the flag is probably not very useful to users who aren't familiar with Java, and there's no guarantee that this log line will remain in sync with changes to the field name or type. This could be clearer with something like:
```suggestion
                log.debug("Performing allocation-balancing revocation immediately as no revocations took place during the previous rebalance");
```

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:
```diff
@@ -307,9 +309,31 @@ ClusterAssignment performTaskAssignment(
+            if (revokedInPrevious && !toExplicitlyRevoke.isEmpty()) {
+                numSuccessiveRevokingRebalances++;
+                delay = (int) consecutiveRevokingRebalancesBackoff.backoff(numSuccessiveRevokingRebalances);
+                log.debug("Consecutive revoking rebalances observed. Need to wait for {} ms", delay);
```
This isn't actually true, is it? We'll need to wait `delay` ms to perform the _next_ allocation-balancing revocation (if another one is necessary), but we're performing the revocation for the current round immediately.

This also highlights a larger issue with the approach here: we should only issue an assignment with a delay when that delay is the time until the re-allocation of lost assignments and/or the next round where we will permit allocation-balancing revocations. With this approach, we'll perform allocation-balancing revocations immediately, but also issue an assignment with a delay, which doesn't really serve much purpose since workers automatically rejoin the group (triggering a rebalance) immediately after every round that included a revocation (see [here](https://github.com/apache/kafka/blob/6f4778301b1fcac1e2750cc697043d674eaa230d/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java#L1567-L1572)).

I think a better approach would be to recompute the potential backoff delay between consecutive allocation-balancing revocations if `delay == 0`, and if it is non-zero, then skip those revocations during the current round.

vamossagar12 replied:
@C0urante , the log line is definitely incorrect and I would change it. Thanks for pointing it out. I had a question on what you proposed as the better approach though.
With the approach I tried to use here, I am allowing `this` revoking rebalance to happen and set a delay so that if another worker joins within that delay, it would need to wait for that much of time before we can perform any revocations. This way the approach is slightly optimistic in the sense atleast the first time around we would have a balanced allocation but further allocation-balancing revocations won't allow a barrage of rebalances in succession.

With what you proposed, it seems to me that even the first such worker joining would be made to wait so that in my mind seems slightly pessimistic. IMO since we are already setting the delay for future consecutive allocation-balancing revocations, it might be better to allow the first one go through. WDYT? Have I understood the context correctly or is there anything that I am missing?

C0urante replied:
By gating the delay behind `revokedInPrevious`, we're always allowing the first round that would require allocation-balancing revocations to actually include those revocations. The proposal I made should only cause us to introduce the delay (and gate revocations behind that delay) after the second consecutive rebalance that would require allocation-balancing revocations.

I'd also like to reiterate this point:
> we should only issue an assignment with a delay when that delay is the time until the re-allocation of lost assignments and/or the next round where we will permit allocation-balancing revocations.

I agree that if we feel comfortable permitting revocations in a given round, then we should just go ahead and include those revocations in the assignment--but in that case, we don't need to include a delay in the assignment, especially since workers will all automatically rejoin the cluster after they've revoked the appropriate connectors/tasks.

yashmayya replied:
Yeah, +1 to @C0urante's point that scheduling a rebalance here doesn't seem to serve any purpose?

> I think a better approach would be to recompute the potential backoff delay between consecutive allocation-balancing revocations if delay == 0, and if it is non-zero, then skip those revocations during the current round.

In this case, we wouldn't know how much time has passed between the previous round of revoking rebalance and the current rebalance right? Or do we not account for that with the assumption that members will rejoin with a new round of rebalance immediately after a revoking round of rebalance anyway?

vamossagar12 replied:
```
Or do we not account for that with the assumption that members will rejoin with a new round of rebalance immediately after a revoking round of rebalance anyway
```

I guess we go by this assumption.

vamossagar12 replied:
Thanks @C0urante , @yashmayya I made the change as suggested. I had made missed the fact that since the workers would join immediately post a load-balancing rebalance, introducing delay for that round is needless. Plz review.

C0urante reviewed (commented):
Thanks @vamossagar12, this is looking much better! Still have a few comments but this is definitely going in the right direction.

vamossagar12 committed:
Update connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java

Co-authored-by: Chris Egerton <fearthecellos@gmail.com>

vamossagar12 commented:
Thanks @C0urante . I had one question on the proposed approach regarding delays. Since the other changes are smallish in nature, I would wait for your response on that one before making the rest of the changes.

Commit:
Review changes

yashmayya commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:
```diff
@@ -285,18 +292,13 @@ ClusterAssignment performTaskAssignment(
-                    performTaskRevocation(activeAssignments, currentWorkerAssignment);
-
-            log.debug("Connector and task to revoke assignments: {}", toRevoke);
+                    performTaskRevocation(configured, completeWorkerAssignment);
```
Why are we changing the first parameter here from `activeAssignments` -> `configured`? `configured` is basically the latest set of tasks and connectors as per the leader's view of the config topic, whereas `activeAssignments` is the assimilation of tasks and connectors from the member assignments of the group in the current round of rebalancing. Is this so that we don't take into account deleted connectors/tasks while doing revocation? This will also take into account new connectors/tasks as well as connectors/tasks lost from the last assignment while calculating the average connector/task load per worker for performing revocations - is this intentional? This seems reasonable to do since if there's no scheduled delayed rebalance, the new as well as lost connector/tasks will be assigned to the workers in this round of rebalance anyway. But I'm just wondering if there's a reason why we were doing revocations only based on existing member assignments earlier.

vamossagar12 replied:
Yes that's exactly the point. I got this logic from the original PR from @showuon  here => https://github.com/apache/kafka/pull/10367 . To quote =>

```
we also passed the configured assignment into performTaskRevocation instead of activeAssignment, so that we can compute the correct expected max assignment number (totalSize / workerSize), instead of the activeTotalSize / workerSize, because the activeTotalSize doesn't include the newAssignments, which will cause the wrong computation, and cause uneven rebalance, or need more round of revoking rebalance.
```

I also noticed that if I don't pass configured, then post allowed + successive revoking rebalance, the assignments were  not balanced.

vamossagar12 replied:
``` 
But I'm just wondering if there's a reason why we were doing revocations only based on existing member assignments earlier.
```
Regarding this I dont know tbh, but my guess is since earlier we were not doing successive revocations, it was not needed. @C0urante can certainly throw some more light on this.

showuon replied:
@yashmayya , yes, your analysis is correct, I was trying to take the new added/deleted connectors/tasks into account for the revocation calculation, because I was trying to do consecutive revocation without delay.

> But I'm just wondering if there's a reason why we were doing revocations only based on existing member assignments earlier.

I guess that's because we didn't consider too complicated scenarios. Anyway, I think this change makes sense.

yashmayya replied:
Thanks @vamossagar12 and @showuon, I do agree that this change makes sense!

C0urante replied:
Switching from `configured` to `activeAssignments` helps in some scenarios, but not others. I found a few cases to demonstrate this while reviewing https://github.com/apache/kafka/pull/10367, documented [here](https://github.com/apache/kafka/pull/10367#discussion_r810393411).

I wonder if we can leave this change out for now and then address the remaining issues with incremental rebalancing in a follow-up PR? There are a lot of edge cases to consider and I don't want to block the primary change here of allowing consecutive revocations (which is super useful on its own!) by tying it to thinking through all the implications for a change that can be made independently.

@showuon what are your thoughts?

vamossagar12 replied:
@C0urante  When you say `I wonder if we can leave this change out for now` do you mean retaining this or switching back to using `activeAssignments` instead of `configured`? I will also take a look at the scenario you described about where this might fail. If we want to retain this, I can take a look at them in a separate PR or this one, whatever is decided over here or if we want to switch back to `activeAssignments`, then that would need more work probably in this PR.

C0urante replied:
I mean switching back to using `activeAssignments`.

vamossagar12 replied:
@showuon let me know what you think about this

showuon replied:
I'm good to revert back to `activeAssignments` and discuss in other PR or JIRA. After all, it's not relevant to this issue. Thanks

vamossagar12 replied:
Thanks @showuon I reverted it back.

Commit:
Setting initial interval to 1 and lowering multiplier. Also always setting delay to 0 to maxDelay is 0.

Commit:
Removing needless check while computing exponential backoff delay

showuon commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:
```diff
@@ -286,30 +294,43 @@ ClusterAssignment performTaskAssignment(
+                    log.debug("Skipping revocations in the current round with a delay of {}ms. Next scheduled rebalance:{}",
+                            delay, scheduledRebalance);
+                } else {
+                    log.debug("Revoking assignments as scheduled.rebalance.max.delay.ms is set to 0");
```
maybe: `Revoking assignments immediately since scheduled.rebalance.max.delay.ms is set to 0`

showuon reviewed (commented):
The changes look good. Left some comments. Also, could you re-enable the test in `RebalanceSourceConnectorsIntegrationTest#testMultipleWorkersRejoining`? This fix should resolve the flaky test.

showuon commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:
```diff
@@ -520,6 +554,9 @@ protected void handleLostAssignments(ConnectorsAndTasks lostAssignments,
             resetDelay();
+            // Resetting the flag as now we can permit successive revoking rebalances.
+            // since we have gone through the full rebalance delay
+            revokedInPrevious = false;
```
Should we also reset `numSuccessiveRevokingRebalances = 0;`? Otherwise, after another successive revoking, the exponential backoff will increase immediately, right?

vamossagar12 replied:
I did think about it.. Problem is if we set the counter to 0 here every time then the exponential backoff would only go upto initial interval after every successive revoking rebalance. The other reason I didn't reset this is that since max interval is maxDelay, this value can be incremented w/o introducing delays > maxDelay. One thing to consider could be can this value overflow if we let it increase unbounded. Let me know what you think about it.

showuon replied:
OK, I got your point. The result will be something like this, right?
1. rebalance: [revokedInPrevious = false, numSuccessive = 0] -> revoke
2. rebalance: [revokedInPrevious = true, numSuccessive = 1] -> delay n 
3. after n (clear revokedInPrevious)
4. rebalance: [revokedInPrevious = false, numSuccessive = 1] -> revoke
5. rebalance: [revokedInPrevious = true, numSuccessive = 1] -> reset numSuccessive or delay

vamossagar12 replied:
Yeah so point `#5`, with the approach I have would set numSuccessive = 2 and calculate backoff delay based on that. If I reset numSuccessive to 0 in `#3` from above, then at `#5`, it would be back to `rebalance: [revokedInPrevious = true, numSuccessive = 1] ` and delay would be equal to the delay we had at `#2` above. 

Basically every successive revoking rebalance would have the same delay because numSuccessive would always be 1 as it's reset and incremented only once. Let me know if that makes sense.

showuon replied:
Yes, make sense to me, to not reseting numSuccessiveRevokingRebalances here. Thanks.

showuon commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java:
```diff
@@ -118,6 +118,152 @@ public void testTaskAssignmentWhenWorkerJoins() {
+    public void testAssignmentsWhenWorkersJoinAfterRevocations()  {
+
+        addNewConnector("connector3", 4);
+        // First assignment with 1 worker and 2 connectors configured but not yet assigned
```
3 connectors?

showuon commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java:
```diff
@@ -118,6 +118,152 @@ public void testTaskAssignmentWhenWorkerJoins() {
+        initAssignor();
+
+        addNewConnector("connector3", 4);
+        // First assignment with 1 worker and 2 connectors configured but not yet assigned
```
3 connectors?

showuon commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java:
```diff
@@ -118,6 +118,152 @@ public void testTaskAssignmentWhenWorkerJoins() {
+        assertConnectorAllocations(0, 1, 2);
+        assertTaskAllocations(3, 3, 6);
+
+        // Follow up rebalance post revocations. No revocations should have happened
+        performStandardRebalance();
+        assertWorkers("worker1", "worker2", "worker3");
+        assertConnectorAllocations(0, 1, 2);
+        assertTaskAllocations(3, 3, 6);
```
Why is there the follow-up rebalance? We don't have revocation happened in previous round, do we?

vamossagar12 replied:
We didn't. I just added it to show even if there's a rebalance it won't cause any revocation. It's unrealistic so to speak but thought would be useful to test. I can remove it if you think it doesn't make sense.

showuon replied:
I think we should remove it. Thanks.

vamossagar12 replied:
Thank you. Removed it.

showuon commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java:
```diff
@@ -118,6 +118,152 @@ public void testTaskAssignmentWhenWorkerJoins() {
+        initAssignor();
+
+        addNewConnector("connector3", 4);
+        // First assignment with 1 worker and 2 connectors configured but not yet assigned
```
3 connectors?

showuon commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java:
```diff
@@ -843,12 +1002,6 @@ public void testTaskAssignmentWhenTasksDuplicatedInWorkerAssignment() {
         assertConnectorAllocations(1, 1);
-        assertTaskAllocations(2, 4);
-
-        // fourth rebalance after revocations
```
The comment in L1008 should change to 
// `Fourth` rebalance should not change assignments

showuon commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java:
```diff
@@ -877,15 +1030,9 @@ public void testDuplicatedAssignmentHandleWhenTheDuplicatedAssignmentsDeleted()
         assertTaskAllocations(0, 2);
 
-        // fourth rebalance after revocations
+        // Third assignment after revocations
```
The comment in L1042 should change to 
// Fourth rebalance should not change assignments

Commit:
Reenabling flaky test testMultipleWorkersRejoining and addressing review comments

vamossagar12 commented:
> The changes look good. Left some comments. Also, could you re-enable the test in `RebalanceSourceConnectorsIntegrationTest#testMultipleWorkersRejoining`? This fix should resolve the flaky test.

Thanks @showuon . I enabled it and it passed locally. Have pushed the change along with other review comments(and a couple of clarifying questions).

Commit:
Removing unwanted rebalance check in testSuccessiveRevocationsWhenMaxDelayIsEqualToExpBackOffInitialInterval

showuon reviewed (approved):
LGTM! Thanks for the improvement!

vamossagar12 commented:
> LGTM! Thanks for the improvement!

Thank you !

Commit:
Changing arguements of performTaskRevocation back to the old ones

vamossagar12 commented:
@C0urante I made the switch back as suggested. Plz review whenever you get the chance.

C0urante commented:
@vamossagar12 there are failing unit tests, can you check those out and ping us after you've taken a look and fixed any issues?

vamossagar12 commented:
> @vamossagar12 there are failing unit tests, can you check those out and ping us after you've taken a look and fixed any issues?

Oops.. Sorry about that. I didn't check the test results after the latest changes. Will check/fix those and call for review post that.

Commit:
Fixing failing testcase

Commit:
Removing unwanted file

Commit:
Increasing exp backoff multiplier to 40

Commit:
Removing resetting numSuccessiveRevokingRebalances to 0

vamossagar12 commented:
> Details

I checked in the changes by fixing the 2 unit tests that were failing. This run, there didn't seem to be connect specific failures. My bad on the oversight the last time around.

C0urante commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorIncrementalTest.java:
```diff
@@ -517,13 +517,13 @@ public void testTaskAssignmentWhenWorkerBounces() {
         assertAssignment(leaderId, offset,
                 Collections.emptyList(), 0,
-                Collections.emptyList(), 0,
+                Collections.emptyList(), 1,
```
Isn't this a regression? We shouldn't be revoking tasks in this round since, without those revocations, we'd have a balanced assignment.

vamossagar12 replied:
Yeah that's true. Revoking tasks at this point leads to imbalance. This was happening since as per the new changes, the moment the delay expires we were allowing revocation. So, at this point, the flag is true, the delay goes to 0 and since revocation is technically possible at this point, the code was doing it. I have added the `canRevoke` flag back to handle this case with which this testcase another test `IncrementalCooperativeAssignorTest#testTaskAssignmentWhenWorkerBounces` which seemed to have a similar issue seem to be fixed.

C0urante replied:
👍 Thanks. I wish we could do this without adding an extra boolean flag but preventing regressions seems worth the additional complexity, especially since there are other issues with the allocation algorithm that need to be worked out and, when fixed, will hopefully render the flag unnecessary.

C0urante replied:
Are there any cases where the introduction of the `canRevoke` flag will defeat the purpose of this PR, by causing an imbalanced assignment to be sent out with no guaranteed follow-up rebalances (such as those caused by a scheduled rebalance delay, or workers rejoining the cluster after receiving an assignment with revocations)?

vamossagar12 replied:
TBH, even I didn't want to re-introduce the flag back but seemed the easiest way to get around the regression. I guess, as you said it might be easier to work through other issues on the allocation algorithm to finally have the flag redundant.

Regarding the side-effects of the re-introduction of this flag, I had imagined that adding the flag back would break some of the tests but that didn't happen which may or mayn't be a good thing. I did look at the logic again and compared with the original algorithm it seemed to me that this line:

https://github.com/apache/kafka/blob/trunk/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java#L310

is the line that prevented successive revoking rebalances. The other check here: https://github.com/apache/kafka/blob/trunk/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java#L312 

gives us a window to set it to true and force a revocation in the next round which kind of made me believe that it should be a safe check. That said, if there are scenarios where we think we need testing, I would be happy to do that.

C0urante replied:
Yep, I was able to produce a test case that causes an imbalanced assignment to be sent out with no revocations or delay by adding a new worker in the middle of a delayed rebalance:

```java
@Test
public void testWorkerJoiningDuringDelayedRebalance() {
    time = new MockTime();
    initAssignor();

    addNewConnector("connector3", 4);
    // First assignment with 1 worker and 3 connectors configured but not yet assigned
    performStandardRebalance();
    assertDelay(0);
    assertWorkers("worker1");
    assertConnectorAllocations(3);
    assertTaskAllocations(12);
    assertBalancedAndCompleteAllocation();

    // Second assignment with a second worker joining and all connectors running on previous worker
    // We should revoke.
    addNewEmptyWorkers("worker2");
    performStandardRebalance();
    assertWorkers("worker1", "worker2");
    assertConnectorAllocations(0, 2);
    assertTaskAllocations(0, 6);

    // Third assignment immediately after revocations, and a third worker joining.
    // This is a successive revoking rebalance. We should not perform any revocations
    // in this round
    addNewEmptyWorkers("worker3");
    performStandardRebalance();
    assertTrue(assignor.delay > 0);
    assertWorkers("worker1", "worker2", "worker3");
    assertConnectorAllocations(0, 1, 2);
    assertTaskAllocations(3, 3, 6);

    // Fourth assignment and a fourth worker joining
    // while delayed rebalance is active. We should not revoke
    time.sleep(assignor.delay / 2);
    addNewEmptyWorkers("worker4");
    performStandardRebalance();
    assertWorkers("worker1", "worker2", "worker3", "worker4");
    assertConnectorAllocations(0, 0, 1, 2);
    assertTaskAllocations(0, 3, 3, 6);

    // Fifth assignment and a fifth worker joining
    // after the delay has expired. We should revoke, but we don't
    time.sleep(assignor.delay);
    addNewEmptyWorkers("worker5");
    performStandardRebalance();
    assertWorkers("worker1", "worker2", "worker3", "worker4", "worker5");
    assertNoRevocations();
    assertDelay(0);
    assertBalancedAndCompleteAllocation();
}
```

It's also worth noting that many of the test cases in this PR need to be updated with this bit at the beginning in order to mock the `Time` instance used by the assignor:
```java
time = new MockTime();
initAssignor();
```
This should be added in any test that invokes `time::sleep`.

At this point, I think we may want to split this into two separate PRs that get merged together. We can revert the `canRevoke` flag from this one, and then add a downstream PR that fixes how we calculate task-balancing revocations in tricky situations like when lost or newly-created tasks have just been assigned. That should fully address cases like the one tested [here](https://github.com/apache/kafka/blob/cda5da9b65f78b51cdfe5371f712a0d392dbdb4d/connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorIncrementalTest.java#L427).

@showuon WDYT? I could extract some of the changes from https://github.com/apache/kafka/pull/12019 that would fix the load-balancing revocation logic and rebase them onto this change if it helps.

showuon replied:
Wow, nice catch!

> At this point, I think we may want to split this into two separate PRs that get merged together. We can revert the canRevoke flag from this one, and then add a downstream PR that fixes how we calculate task-balancing revocations in tricky situations like when lost or newly-created tasks have just been assigned. That should fully address cases like the one tested [here](https://github.com/apache/kafka/blob/cda5da9b65f78b51cdfe5371f712a0d392dbdb4d/connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinatorIncrementalTest.java#L427).

Agree! Let's revert the canRevoke flag and then deal with tricky cases in a follow-up PR. Thank you.

vamossagar12 replied:
Thanks Chris/Luke. I have revoked the `canRevoke` flag commit.  Also, Chris, I looked at your PR and it's closed but not merged. I see some changes specially =>
```
Perform load-balancing revocations any time the cluster appears imbalanced and there are still connectors and tasks that can be revoked from workers, instead of only when the number of workers in the cluster changes
```
which sounds interesting (and remember discussing this on the ticket for this). Also, let me know if there's anything I can also contribute to for the other PR (Looking at your comment it appeared to me that you would be rebasing the commits from that closed PR and hence the question).

vamossagar12 replied:
hey Chris.. Sorry to bug you again on this one. Since we anyways won't be merging this PR w/o the second one, would it make sense to incorporate the changes to this one? I believe the changes are already there in the other PR of yours?

vamossagar12 replied:
> Isn't this a regression? We shouldn't be revoking tasks in this round since, without those revocations, we'd have a balanced assignment.

Hello.. I have been playing around with the revokedInPrevious and the test cases. I looked at this case which is a regression which I agree. However, based on the changes made in this PR, what I think is that post the revoking rebalance, there would be a follow up rebalance which would eventually lead to a balanced load. If you look at `IncrementalCooperativeAssignorTest#testTaskAssignmentWhenWorkerBounces` this behaviour  is exhibited. On similar lines, I made some edits to `WorkerCoordinatorIncrementalTest#testTaskAssignmentWhenWorkerBounces` method by causing a follow up rebalance after the revoking rebalance after worker 3 comes back. The assignments with which I am calling the final `onLeaderElected` may not be accurate (as I couldn't get the full assignments using the deserialzeAssignment) but I just chose something which is representative.

I know this is a deviation from what the KIP proposes i.e a bounced member gets its assignments back without any revocations but if what I am stating above sounds ok, do you think it's really bad- coz it seems to be getting to a  balanced load after a a follow up rebalance. WDYT?

C0urante replied:
Hey @vamossagar12--I'd rather err on the side of not increasing the number of rebalances required for simple cases like the one covered in `WorkerCoordinatorIncrementalTest::testTaskAssignmentWhenWorkerBounces`.

I plan to file the downstream PR sometime next week. For ease of review I'd prefer to keep it separate from this one, which has grown fairly large and complex on its own.

vamossagar12 replied:
Thanks @C0urante . I have reverted the test with the follow up rebalance I was basing my thoughts on the fact that KIP-415 allows having more rebalances so adding another one should be ok. Having said that, I am fine with whatever you stated as well.

Regarding a separate PR, yes that makes sense. Thanks for your help on this!

C0urante replied:
Hi @vamossagar12--sorry for the delay. I've filed the follow-up PR here: https://github.com/vamossagar12/kafka/pull/1

I've marked it as a draft since it should not be merged into your KAFKA-12495 branch until it's approved to be merged into trunk here, but it's ready for review now. Can you take a look and see what you think? If it looks alright, we can ask Luke to take a look.

(cc @showuon)

vamossagar12 replied:
Thanks @C0urante ! I have added some comments for some understanding and around naming.

Commit:
Resetting revokedInPrevious to false when there's no rebalance + few tests moved around/fixed

Commit:
Removing unwanted sysout

Commit:
Reverting test

C0urante committed:
KAFKA-12495: Improve rebalance allocation algorithm

C0urante committed:
Address review comments, remove unused code

C0urante commented:
@showuon The follow-up PR [here](https://github.com/vamossagar12/kafka/pull/1) is ready for review; this should be the last step necessary before we can merge this PR and fix this long-standing rebalancing bug.

vamossagar12 commented:
@showuon Whenever you get the chance, can you also plz review [this](https://github.com/vamossagar12/kafka/pull/1) PR  of Chris. We should be able to get this one through as well.

C0urante committed:
Address review comments

showuon commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/RebalanceSourceConnectorsIntegrationTest.java:
```diff
@@ -270,7 +270,6 @@ public void testRemovingWorker() throws Exception {
                 WORKER_SETUP_DURATION_MS, "Connect and tasks are imbalanced between the workers.");
     }
 
-    @Ignore // TODO: To be re-enabled once we can make it less flaky (KAFKA-12495, KAFKA-12283)
```
@vamossagar12 , in L191, there's also a disabled test, I think after this patch, they should be reliable. Please enable it, too.

vamossagar12 replied:
done.

Commit:
Reenabling RebalanceSourceConnectorsIntegrationTest#testDeleteConnector test

vamossagar12 committed:
Merge pull request #1 from C0urante/kafka-12495-followup

KAFKA-12495: Improve rebalance allocation algorithm

vamossagar12 commented:
@C0urante , you want to take a final pass on this one whenever you get the chance?

Commit:
Fixing checkstyle

Commit:
Revert "Fixing checkstyle"

This reverts commit 2eb1f3fbedb793353d58f61a6a0074b9bbfb6063.

vamossagar12 commented:
@C0urante , i did that but looks like my force push cleaned up the PR merge since I didn't pull the latest :( I see you have deleted the branch. Would it be possible to create another PR for this? Sorry about this :(

Commit:
fixing checkstyle again

C0urante commented:
@vamossagar12 It looks like the commit is still present in GitHub: https://github.com/vamossagar12/kafka/tree/422c81fc0a2dd19fee0ed1eb033ef35fc9e27ba1

Can you fetch that commit locally, then add a commit on top of it that removes the `@Ignore` annotation (and the import for it), and push that to this branch?

C0urante commented:
Thanks @vamossagar12. I've run the tests locally and they've succeeded; will wait to see what happens with CI but I think this should be okay.

C0urante reviewed (approved):
LGTM, thanks Sagar for the PR and Luke for the additional review. Great to finally get a fix in for this long-standing issue!

C0urante commented:
The single test failure is unrelated. Merging...

showuon commented:
One question to @C0urante , do you think we should backport to v3.3 branch?

C0urante commented:
@showuon I was thinking about it--IMO the additional complexity here is a little too risky to backport but I'm almost on the fence here. If you'd like to see it on 3.3 I wouldn't be opposed

vamossagar12 commented:
I think 3.4 release is also around the corner(few weeks maybe?). Would it better to have it in 3.4?

showuon commented:
Yes, I have the same thought as Chris. Besides, this bug already exist for a long long time, it's not that urgent to put it into 3.3. So, let's keep it in 3.4 (trunk branch), which is what we currently did. :)

vamossagar12 commented:
hwy @C0urante , I was thinking should the exponential backoff thing that we have introduced as part of this PR should go somewhere in the docs? I am saying this since this is a deviation from how things worked prior to this. WDYT?

C0urante commented:
@vamossagar12 I don't think it's necessary to call this out anywhere unless it's caused unexpected issues with our users. The intention behind the exponential backoff is to avoid rebalance storms but I'm still not sure when those would ever realistically happen, so until we see otherwise, I'd prefer to leave it as an internal implementation detail.
